database.c

```diff
@@ -728,6 +737,65 @@ database_save_later(int secs)
   }
 }
 
+void
+migrate_configs(const char *legacydir, const char *currentdir)
+{
+  DIR *dptr;
+  struct dirent *dir;
+  FILE *src, *dest;
+  char srcname[260], destname[260];
+  int ch;
+  errno = 0;
+  if (rename(legacydir, currentdir) < 0)
+  {
+    log_warning("Rename failed with %s. Falling back to copy/delete.\n", strerror(errno));
+    if (mkdir(currentdir, 0755) < 0 && errno != EEXIST)
+    {
+      log_err("%s: mkdir: %s\n", currentdir, strerror(errno));
+      return;
+    }
+    dptr = opendir(legacydir);
+    if (dptr)
+    {
+      while ((dir = readdir(dptr)) != NULL)
+      {
+        if (dir->d_type == DT_REG)
+        {
+          snprintf(srcname, sizeof(srcname), "%s/%s", legacydir, dir->d_name);
+          snprintf(destname, sizeof(destname), "%s/%s", currentdir, dir->d_name);
+          src = fopen(srcname, "r");
+          dest = fopen(destname, "w");
+          if (src == NULL || dest == NULL)
+          {
+            log_err("File open failed: %s %s\n", srcname, destname);
+            return;
+          }
+          while((ch = fgetc(src)) != EOF)
+          {
+            fputc(ch, dest);
+          }
+          fclose(src);
+          fclose(dest);
+          if (remove(srcname) != 0)
+          {
+            log_warning("Failed to remove %s with %s\n", srcname, strerror(errno));
+          }
+        }
+      }
+    }
+    else
+    {
+      log_err("Directory open failed: %s\n", legacydir);
+      return;
```

Comment:
In generic, for readability: Handle errors/unexpected things first, ie. in if-block, not else-block

Also, chances are that directory could be removed even if it can't be read, so... might try it instead of bailing out?

Reply:
In such cases a goto fail also could work quite nicely.